### PR TITLE
Adding profile support for registry 

### LINKED ISSUE #435: Add support for multiple AWS credential profiles
**Is your feature request related to a problem? Please describe.**

Currently, OADP Operator does not support multiple AWS credential profiles for the secret to be used for BSL/VSL/Registry.

**Describe the solution you'd like**

We would like to add support for multiple AWS profiles. This might require AWS secret parsing changes so that the secret data is appropriately consumed by the Registry environment vars.

**Additional context**

Reference: https://github.com/openshift/oadp-operator/blob/0a6008d2a5e3603827727d7b57904b4040540325/controllers/registry.go#L505

### PR DESCRIPTION
**Pull request by deepakraj1997**
Resolves #435

### CODE REVIEW

File: controllers/registry.go [hunk 1]
```diff
@@ -37,6 +37,7 @@ const (
 	RegistryStorageS3AccesskeyEnvVarKey      = "REGISTRY_STORAGE_S3_ACCESSKEY"
 	RegistryStorageS3BucketEnvVarKey         = "REGISTRY_STORAGE_S3_BUCKET"
 	RegistryStorageS3RegionEnvVarKey         = "REGISTRY_STORAGE_S3_REGION"
+	RegistryStorageS3ProfileEnvVarKey        = "REGISTRY_STORAGE_S3_PROFILE"
```
Comment (deepakraj1997) [inline controllers/registry.go:40]: Added this for consistency. Not really required.

Comment (shubham-pampattiwar) [reply controllers/registry.go:40]: @deepakraj1997 As this env var is not actually required or consumed by the registry deployment can we get rid of it ? IMO It's presence might give a false sense that it is being used somewhere for configuration purposes.

File: controllers/registry.go [hunk 2]
```diff
@@ -383,9 +389,12 @@ func (r *VeleroReconciler) getAWSRegistryEnvVars(bsl *velerov1.BackupStorageLoca
 		r.Log.Info(fmt.Sprintf("Error fetching provider secret %s for backupstoragelocation %s/%s", secretName, bsl.Namespace, bsl.Name))
 		return nil, err
 	}
-
+	awsProfile := "default"
+	if value, exists := bsl.Spec.Config[Profile]; exists {
```
Comment (shubham-pampattiwar) [inline controllers/registry.go:393]: +1, nice use of the Config spec map.

File: controllers/registry.go [hunk 3]
```diff
@@ -502,7 +519,7 @@ func (r *VeleroReconciler) getSecretNameAndKey(credential *corev1.SecretKeySelec
 	return secretName, secretKey
 }
 
-func (r *VeleroReconciler) parseAWSSecret(secret corev1.Secret, secretKey string) (string, string, error) {
+func (r *VeleroReconciler) parseAWSSecret(secret corev1.Secret, secretKey string, awsProfile string) (string, string, error) {
 
 	AWSAccessKey, AWSSecretKey := "", ""
 	// this logic only supports single profile presence in the aws credentials file
```
Comment (shubham-pampattiwar) [inline controllers/registry.go:525]: Please remove this comment, as this PR adds support for multiple profiles.

File: controllers/registry.go [hunk 4]
```diff
@@ -519,44 +536,59 @@ func (r *VeleroReconciler) parseAWSSecret(secret corev1.Secret, secretKey string
...
-				return AWSAccessKey, AWSSecretKey, errors.New("secret parsing error")
+			awsProfileRegex, err := regexp.Compile(`\[|\]`)
+			if err != nil {
+				return AWSAccessKey, AWSSecretKey, errors.New("parseAWSSecret faulty regex: keyNameRegex")
 			}
-			AWSAccessKey = splitLine[1]
-			continue
-		}
+			cleanedLine := strings.ReplaceAll(line, " ", "")
+			parsedProfile := awsProfileRegex.ReplaceAllString(cleanedLine, "")
```
Comment (shubham-pampattiwar) [inline controllers/registry.go:549]: @deepakraj1997 Can you please add a log for the case when the supplied aws profile is not found in the parsing of the credential secret ? like the config had the name `[test-profile]` but this profile was not to be found while parsing then this log might come in handy. I was thinking something like 
`if (index +1) == len(splitString) { 
errors.New("Supplied profile not found in parsing...") some thing similar 
}
`

**Review by shubham-pampattiwar — CHANGES_REQUESTED** [review]
@deepakraj1997 Thank you for the PR. Overall changes look great ! Just added some suggestions.

File: controllers/registry.go (hunk 1 above)
Comment (deepakraj1997) [reply controllers/registry.go]: Thanks for reviewing. Removed.

File: controllers/registry.go (hunk 2 above)
Comment (deepakraj1997) [reply controllers/registry.go:388]: Removed the env vars.

File: controllers/registry.go (hunk 4 above)
Comment (deepakraj1997) [reply controllers/registry.go:535]: Hey shubham, I think if we do that we wouldnt know where the exact error is (Like if its access key/secret missing) and the lines from 585-592 will become dead code.  Maybe I am overthinking about the edge cases...I have added an alternative check to pinpoint where the error would be.

File: controllers/registry.go (hunk 3 above)
Comment (deepakraj1997) [reply controllers/registry.go]: Removed.

**Comment by deepakraj1997** [comment]
@shubham-pampattiwar thanks for the review. I have made the requested changes. Do review when u have time. Thanks!

File: controllers/registry.go (hunk 4 above)
Comment (shubham-pampattiwar) [reply controllers/registry.go:535]: @deepakraj1997 yeah makes sense, the alternative check works. Thanks !

File: controllers/registry.go [hunk 5]
```diff
@@ -502,10 +506,9 @@ func (r *VeleroReconciler) getSecretNameAndKey(credential *corev1.SecretKeySelec
...
 }
 
-func (r *VeleroReconciler) parseAWSSecret(secret corev1.Secret, secretKey string) (string, string, error) {
+func (r *VeleroReconciler) parseAWSSecret(secret corev1.Secret, secretKey string, matchProfile string) (string, string, error) {
 
-	AWSAccessKey, AWSSecretKey := "", ""
-	// this logic only supports single profile presence in the aws credentials file
+	AWSAccessKey, AWSSecretKey, AWSProfile := "", "", ""
 	splitString := strings.Split(string(secret.Data[secretKey]), "\n")
 	keyNameRegex, err := regexp.Compile(`\[.*\]`) //ignore lines such as [default]
```
Comment (shubham-pampattiwar) [inline controllers/registry.go:513]: @deepakraj1997  NIT: The existing comment for this regex utility is misleading. Can you please remove/update this comment ?

File: controllers/registry.go [hunk 6]
```diff
@@ -519,46 +522,66 @@ func (r *VeleroReconciler) parseAWSSecret(secret corev1.Secret, secretKey string
...
+						AWSSecretKey = splitLine[1]
+						continue
+					}
+				}
 			}
-			AWSSecretKey = splitLine[1]
-			continue
 		}
 	}
+	if AWSProfile == "" {
```
Comment (shubham-pampattiwar) [inline controllers/registry.go:581]: NIT: AWS Profile -> just profile

**Review by shubham-pampattiwar — APPROVED** [review]
Just added a couple of nits. Apart from that LGTM ! Thank you !

File: controllers/registry.go (hunk 5 above)
Comment (deepakraj1997) [reply controllers/registry.go]: Yep, my bad, didnt notice this comment.

File: controllers/registry.go (hunk 6 above)
Comment (deepakraj1997) [reply controllers/registry.go]: changed.

**Review by dymurray — APPROVED** [review]

**Pull request merged**